### PR TITLE
Kos15 improvements

### PR DESCRIPTION
**Pull request by th4s**
Adapt the current KOS15 implementation to make use of matrix-transpose and add some minor improvements

- [x] Get rid of padding for choice bits
- [x] Only allow matrices with dimensions of the form `2^n`
- [x] Matrices should make use of `Vec<T>` instead of `Vec<Vec<T>>`
- [x] Number of extended OTs / choices is no longer necessary on init

### CODE REVIEW

File: mpc-core/src/ot/extension/kos15/receiver.rs [hunk 1]
```diff
@@ -108,14 +110,13 @@ where
     R: Rng + CryptoRng,
     C: BlockCipher<BlockSize = U16> + BlockEncrypt,
 {
-    pub fn new_with_custom(mut rng: R, cipher: C, base: BaseSender, count: usize) -> Self {
+    pub fn new_with_custom(mut rng: R, cipher: C, base: BaseSender) -> Self {
```
[the remark this answers is not in the document]
Comment (th4s) [reply mpc-core/src/ot/extension/kos15/receiver.rs:113]: yup

File: mpc-core/src/ot/extension/kos15/receiver.rs [hunk 2]
```diff
@@ -204,58 +205,56 @@ where
     ) -> Result<ExtReceiverSetup, ExtReceiverCoreError> {
         check_state(&self.state, &State::BaseSend)?;
 
+        // For performance purposes and to get rid of padding we require that choice is a
+        // multiple of 2^k for some k
+        if choice.len() % 256 != 0 {
```
[the remark this answers is not in the document]
Comment (th4s) [reply mpc-core/src/ot/extension/kos15/receiver.rs]: yepp

File: mpc-core/src/ot/extension/kos15/sender.rs [hunk 3]
```diff
@@ -233,9 +214,14 @@ where
     ) -> Result<(), ExtSenderCoreError> {
         check_state(&self.state, &State::BaseReceive)?;
 
-        let ncols = receiver_setup.table[0].len() * 8;
+        let ncols = receiver_setup.ncols;
+
+        // This is because of the extension of the receiver choices with extra padding bytes
+        // We have to subtract them here.
+        self.count = ncols - receiver_setup.padding;
```
[the remark this answers is not in the document]
Comment (th4s) [reply mpc-core/src/ot/extension/kos15/sender.rs]: yep, done.

**Review by themighty1 — APPROVED** [review]
ACK

**Pull request merged**